File: sources/opencv-samples/15_HistogramEqualizationRGB/main.py
import cv2 as cv
import sys
import argparse
from time import time


def getTime():
	return time()


def getTimeDelta(t):
	t = 1000 * (getTime() - t)
	return round(1000 * t) / 1000


def main():
	args = parser.parse_args()
	
	path = cv.samples.findFile(args.img_path)
	
	if not path:
		sys.exit(f"Image {args.img_path} not found")
	
	image = cv.imread(path)

	if image is None:
		sys.exit(f"Could not read the image {path}")

	window1 = "Image"
	window2 = "Equalized"

	t = getTime()

	# Séparer les canaux de couleur
	b, g, r = cv.split(image)

	# Appliquer l'égalisation d'histogramme à chaque canal de couleur
	b = cv.equalizeHist(b)
	g = cv.equalizeHist(g)
	r = cv.equalizeHist(r)

	# Fusionner les canaux de couleur en une seule image
	equalized = cv.merge((b, g, r))

	print(f"duration: {getTimeDelta(t)} ms")

	cv.imshow(window1, image)
	cv.imshow(window2, equalized)

	cv.waitKey(0)
	cv.destroyAllWindows()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Histogram Equalization RBG main.py CLI")
    parser.add_argument("img_path", type=str, help="image path to open")
    main()

File: sources/opencv-samples/15_HistogramEqualizationRGB/main.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>

using namespace std;
using namespace cv;


double getTime()
{
    return (double)getTickCount();
}

double getTimeDiff(double time_ori)
{
    return 1000 * (getTime() - time_ori) / getTickFrequency();
}

int main(int argc, char** argv) 
{
    if (argc != 2)
    {
        printf("Usage: ./main.out <image_path>");
        return -1;
    }

    Mat image = imread(argv[1]);

    if (!image.data)
    {
        cout << "Could not read the image " << argv[1] << endl;
        return -1;
    }

    string window1 = "Image";
    string window2 = "Equalized";

    double t = getTime();

    // Séparer les canaux de couleur
    vector<Mat> channels;
    split(image, channels);
    Mat b = channels[0];
    Mat g = channels[1];
    Mat r = channels[2];

    // Appliquer l'égalisation d'histogramme à chaque canal de couleur
    equalizeHist(b, b);
    equalizeHist(g, g);
    equalizeHist(r, r);

    // Fusionner les canaux de couleur en une seule image
    Mat equalized;
    merge(vector<Mat>({b, g, r}), equalized);

    cout << "duration : " << getTimeDiff(t) << " ms" << endl;

    imshow(window1, image);
    imshow(window2, equalized);

    waitKey(0);
    destroyAllWindows();
    return 0;
}
